### PR TITLE
Medaka update

### PR DESCRIPTION
**Pull request by NatPRoach**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [X] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

**Comment by bgruening** [comment]
@NatPRoach  1.1.3 is out https://anaconda.org/bioconda/medaka/

**Comment by NatPRoach** [comment]
> @NatPRoach 1.1.3 is out https://anaconda.org/bioconda/medaka/

Yes I will try to get 1.1.3 wrapped shortly. Just trying to fix problems with the current version. I'll try to open a new PR for the new version shortly.

**Comment by NatPRoach** [comment]
Commenting for record keeping purposes RE: @bgruening 's comment that medaka 1.1.3 is released. It would seem that medaka 1.1.3 makes the switch to tensorflow2 and as such many of it's old models are no longer functional in the new version and are not packaged in the conda package. We therefore decided to update v1.0.3 with the new approach to annotating variants first, then update the wrapper to 1.1.3, such that there is a valid version of medaka on tools-iuc that can do variant calling on reads basecalled with older versions of the basecallers.

File: tools/medaka/variant.xml [hunk 1]
```diff
@@ -47,12 +47,16 @@
...
-    && grep -v "^#" '$out_result' > tmp.txt
-    && cut -f 1,2 tmp.txt > positions.txt
-    && samtools mpileup -d 0 --positions positions.txt -Q $output_annotated.min_basequal -q $output_annotated.min_mapqual -f reference.fa '$output_annotated.in_bam' > mpileup.txt
-    && python '$__tool_directory__/annotateVCF.py' '$out_result' mpileup.txt '$out_annotated'
+#if $out_annotated:
+##    && samtools faidx reference.fa
+##    && grep -v "^#" '$out_result' > tmp.txt
+##    && cut -f 1,2 tmp.txt > positions.txt
+##    && samtools mpileup -d 0 --positions positions.txt -Q $output_annotated.min_basequal -q $output_annotated.min_mapqual -f reference.fa '$output_annotated.in_bam' > mpileup.txt
+##    && python '$__tool_directory__/annotateVCF.py' '$out_result' mpileup.txt '$out_annotated'
```
Comment (wm75) [inline tools/medaka/variant.xml:55]: No need to keep these old lines around - remove.

File: tools/medaka/variant.xml [hunk 2]
```diff
@@ -84,10 +88,14 @@
                 <option value="true" selected="true">Output annotated VCF</option>
                 <option value="false">Don't output annotated VCF</option>
             </param>
-            <when value="true">
+            <!-- <when value="true">
                 <param name="in_bam" type="data" format="bam" optional="false" label="BAM to annotate the VCF" />
                 <param name="min_basequal" type="integer" value="0" min="0" max="50" label="Minimum base quality" help="The minimum base quality (default: 0) at a given variant position required to include a read in the calculation of variant statistics (samtools mpileup -Q)" />
                 <param name="min_mapqual" type="integer" value="0" min="0" max="60" label="Minimum mapping quality" help="The minimum mapping quality (default: 0) required for a read to be considered in calculation of variant statistics (samtools mpileup -q)" />
+                <param name="max_depth" type="integer" value="0" min="0" max="2147483647"/>
+            </when> -->
```
Comment (wm75) [inline tools/medaka/variant.xml:96]: remove old code

File: tools/medaka/convert_VCF_info_fields.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+# 10/21/2020 - Nathan P. Roach, natproach@gmail.com
+
+import sys
+from math import log10
+
+from scipy.stats import fisher_exact
+
+
+def pval_to_phredqual(pval):
+    return int(round(-10. * log10(pval)))
```
Comment (wm75) [inline tools/medaka/convert_VCF_info_fields.py:17]:
```suggestion
    return round(-10 * log10(pval))
```

but is also used only once, so could be inlined.

**Review by wm75 — CHANGES_REQUESTED** [review]
I think this PR still contains the now abandoned annotateVCF.py script, which can be removed.

Also note that the convert_VCF_info_fields script would emit now one VCF record per ALT allele, but without splitting the ALT field itself.
AFAIU, medaka will never emit multiallelic records so this won't be an issue then, but since the code tries to be able to deal with such records, it should do so correctly.

**Comment by NatPRoach** [comment]
> Also note that the convert_VCF_info_fields script would emit now one VCF record per ALT allele, but without splitting the ALT field itself.
> AFAIU, medaka will never emit multiallelic records so this won't be an issue then, but since the code tries to be able to deal with such records, it should do so correctly.
You're absolutely right, I'll fix that thanks for pointing it out.

File: tools/medaka/convert_VCF_info_fields.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    for i, line in enumerate(in_vcf):
+        if i == 1:
+            out_vcf.write("##convert_VCF_info_fields=0.1\n")
+        if line[0:2] == "##":
+            if "Number=." in line:
+                assert line[0:11] == "##INFO=<ID="
+                line = line[11:]
+                id_ = line.split(',')[0]
+                to_skip.add(id_)
+                continue
```
[the remark this answers is not in the document]
Comment (NatPRoach) [reply tools/medaka/convert_VCF_info_fields.py:44]: I suppose that makes sense. My logic was that "Number=." fields will probably be made in such a way that they will scale the number of entries based on number of variants, but I suppose that's not always true.

**Comment by bernt-matthias** [comment]
Not sure yet what the problem is over here https://github.com/galaxyproject/tools-iuc/issues/3270 .. but it would be great if it would be solved within this PR :)

**Comment by NatPRoach** [comment]
> Not sure yet what the problem is over here #3270 .. but it would be great if it would be solved within this PR :)

That's the current plan. Apologies will get to addressing #3270 soon, been slammed with other stuff but it's on the list.

**Comment by NatPRoach** [comment]
Oh, the issue mentioned in #3270 , that read groups are ignored in medaka_variant pipeline is not an issue on our end. It's a limitation of medaka v1.0.3 : 
Relevant section of `medaka_variant --help`
```
-i  input bam of reads aligned to ref. Read groups are currently ignored,
        so the bam should only contain reads from a single sample.
```

**Comment by bgruening** [comment]
ping @NatPRoach

**Comment by NatPRoach** [comment]
> ping @NatPRoach

Hi @bgruening, I think the status on this is where this thread left it previously. The changes @wm75 requested have been addressed I believe but there was no movement on merging it because of this issue with the medaka annotate function https://github.com/nanoporetech/medaka/issues/192 .

I personally think we should move ahead with the merge anyway, both because our current approach to annotating variants fails with large genomes or high depth of coverage because of memory issues, and we should address that in a version of the wrapper with the old models based on tensorflow 1, since the version after this drops a bunch of models in the upgrade to tensorflow 2. The turn around to bumping a wrapper to the next version of medaka should be quite short after that.

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
@wm75 are you ok if we proceed here?

**Review by davebx — APPROVED** [review]

**Comment by NatPRoach** [comment]
> @wm75 are you ok if we proceed here?
Ping @wm75 . Would appreciate an up or down decision when you get the chance.

**Review by wm75 — COMMENTED** [review]
Really sorry for not getting round to this again earlier :(

Just tiny comments this time, then everything looks ready to go.

**Review by wm75 — APPROVED** [review]

**Comment by wm75** [comment]
Thank you @NatPRoach !

**Pull request merged**